tests/test_validators.py

```diff
@@ -2779,6 +2781,42 @@ class MyModel(BaseModel):
     assert m.foobar == {'value': '1', 'field_name': 'foobar', 'data': {'x': 123}}
 
 
+@pytest.mark.skipif(sys.version_info < (3, 12), reason='requires Python 3.12+ for PEP 695 type keyword')
+def test_validator_field_name_with_reused_type(create_module):
+    mod = create_module(
+        textwrap.dedent(
+            """
+        from typing import Annotated
+        from pydantic import BaseModel, AfterValidator, ValidationInfo
+
+        calls = []
+
+        def validate_my_field(value: str, info: ValidationInfo):
+            calls.append((info.field_name, value))
+            return value
+
+        type MyField = Annotated[str, AfterValidator(validate_my_field)]
```

Comment:
```suggestion
        MyField =  TypeAliasType('MyField', Annotated[str, AfterValidator(validate_my_field)])
```
You can make use of `typing_extensions.TypeAliasType` so that the test can run on all versions, without the `create_module()` helper.